### PR TITLE
Code review upgrade

### PR DESCRIPTION
**Pull request by sof202**
# Description

This pull request will upgrade the code review section of the website. Previously there were too many things to look for in a code review and a few of them (readability, understandability and maintainability) were poorly described and overlapping. To make code reviews easier to understand, this section was revised. Some bad coding practices due to 'overly clean code' were removed or warned about also.

## Issue ticket number

This pull request is to address issue: #55.

## Type of pull request

- [ ] Bug fix
- [ ] New feature/enhancement
- [ ] Code refactor
- [x] Documentation update

## Checklist

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my code to check that it is functional
- [x] I have used linters to check for common sources of errors
- [ ] I have implemented fail safes in my code to account for edge cases
- [x] I have made the corresponding changes to the documentation

### CODE REVIEW

**Review by SiyiSEA — COMMENTED** [review]
Better version of the explanation!! 
But there are some paragraphs that I can't understand properly.
I've commented them, hopefully.

File: documentation/docs/Developer-information/Code-review/Code-examples.md [hunk 1]
```diff
@@ -10,11 +10,13 @@ When conducting a code review, there are many aspects the reviewer should be loo
 
 ## Design
 
-This is a rather abstract concept and goes beyond the following example, the full description given in the [list](./Conducting-a-code-review.md#what-to-look-for-in-a-code-review) is more general than the example given.
+Suppose the file `main.py` currently utilizes an external package, such as numpy, to source the factorial function. Now, imagine a new pull request introduces a customized version of the factorial function, claiming that it is superior to the original implementation.
 
-Suppose you have the file: `main.py`. At some point in `main.py`, the factorial function is required (for some calculation). Suppose further that `main.py` is currently using an external package (like numpy) to source this function.
+This scenario raises concerns about code design and consistency. If the new factorial function is accepted without considering its impact on other parts of the codebase, inconsistencies may arise. For instance, if the original factorial function is used in multiple files, accepting the new implementation could lead to discrepancies and potential errors in other areas of the codebase.
```
[the remark this answers is not in the document]
Comment (sof202) [reply documentation/docs/Developer-information/Code-review/Code-examples.md]: I've slightly reworded this section to make it more clear why the pull request should be rejected in the laid out scenario. 
I've also added an additional paragraph quickly noting the requirements for accepting the pull request if the customised version of the function was truly necessary.

File: documentation/docs/Developer-information/Code-review/Code-examples.md [hunk 2]
```diff
@@ -187,11 +127,15 @@ def factorial(n):
     return result
 ```
 
-Maintainability is not limited to deeply nested code, this is just one example of how code can be difficult to maintain due to it being poorly written.
+Note that readability/understandability/maintainability can be subjective in lots of scenarios. For now, if you, the reviewer, understands the code to a 'good enough' degree (and you are not completely lost), the code is fine and does not warrant a refactoring. There is [no such thing as perfect code](./Best-practices.md#there-is-no-perfect-code). It is very easy to go back and forth on what makes code cleaner and easier to read, but all this really ends up doing is wasting time. If the proposed changes are not completely unreadable and improves the existing codebase, approve the request.
```
[the remark this answers is not in the document]
Comment (sof202) [reply documentation/docs/Developer-information/Code-review/Code-examples.md:131]: That's my bad. It makes more sense when I read it out loud. Saying simply 'readable' on it's own isn't quite enough as it is quite subjective. I've put 'reasonably readable' instead (which is closer to what I mean by 'not completely unreadable')

File: documentation/docs/Developer-information/Code-review/Code-examples.md [hunk 3]
```diff
@@ -226,11 +170,11 @@ def factorial(n):
...
+Sometimes, explaining *what* the code does feel like a necessity. In such cases we refer back to our stance on [complex code](#complexity). If the code is complex enough that *what*-based comments are required, then the code would likely benefit from being refactored. The exception to this is circumstances like regular expressions. If the code uses something like regex, then the comments should explain what the code is doing. Computers read regex, humans do not (well, not easily).
 
 ## Scalability/Expandability
 
-Can the code be repurposed in a new context? Again, we are going to stick with the factorial function for simplicity, spotting scalability problems is usually rather difficult (without lots of domain knowledge).
+Can the code be repurposed in a new context? We appreciate that spotting scalability problems is usually rather difficult in more complex codebases without lots of domain knowledge. Generally, you should be looking to see if some code can be modularised so that it can be repurposed elsewhere.
```
[the remark this answers is not in the document]
Comment (sof202) [reply documentation/docs/Developer-information/Code-review/Code-examples.md:178]: You raise a good point, I've added an additional example. Scalability is not just about reducing repetitions. The additional example covers the 'single responsibility principle', which states that a function/class should do *one* thing only.

File: documentation/docs/Developer-information/Code-review/Conducting-a-code-review.md [hunk 4]
```diff
@@ -101,61 +101,45 @@ If the pull request had been accepted without a review would anything else break
...
 
-Ideally a script (or source code) can be easily repurposed if required. A tool might work for a specific dataset, but a better tool could work with other datasets with a few small changes. The script (or source code) does not need to be multi-purpose itself, but the ability to be such can be highly beneficial.
-
-### [Style](./Code-examples.md#style)
-
-In general, style only matters for consistency. We encourage that contributors use linters to check for their coding style so this is unlikely to be a big problem. Bad coding style is generally fine unless it results in the code failing one of the above points. In general you are as unlikely to change someone else's coding style as they are to change your own. Style should be about consistency, not about personal preference.
-
-### [Documentation](./Code-examples.md#documentation)
+Ideally, a script (or source code) can be easily repurposed if required. A tool might work for a specific dataset, but a better tool could work with other datasets with a few small changes. The script (or source code) does not need to be multi-purpose itself, but the ability to be such can be highly beneficial.
```
[the remark this answers is not in the document]
Comment (sof202) [reply documentation/docs/Developer-information/Code-review/Conducting-a-code-review.md:138]: I've decided to remove the documentation section from the code review section as I don't want to force everyone to update the documentation with every little change that they make. I'm planning on adding a separate page for the points that you are talking about.

File: documentation/docs/Developer-information/Code-review/Conducting-a-code-review.md [hunk 5]
```diff
@@ -101,61 +101,45 @@ If the pull request had been accepted without a review would anything else break
...
 
-Has the associated documentation been properly updated?
+### [Useful: Style](./Code-examples.md#style)
 
-This is the least important aspect of a pull request. In the event that documentation has been forgotten or incorrectly altered, an issue can be created and documentation can be created properly in the future.
+In general, style only matters for consistency. We encourage that contributors use linters/prettiers to check their code for errors and style so this is unlikely to be a big problem. Bad coding style is generally fine unless it results in the code failing one of the above points. In general you are as unlikely to change someone else's coding style as they are to change your own. Style should be about consistency with the current code base, not about personal preference.
```
[the remark this answers is not in the document]
Comment (sof202) [reply documentation/docs/Developer-information/Code-review/Conducting-a-code-review.md:142]: You're right, I was thinking this when I was writing up this pull request. I've collapsed all of the points into ##On the matter of style. There is a lot of repetition between these sections that isn't needed.

File: documentation/docs/Developer-information/Code-review/Code-examples.md [hunk 6]
```diff
@@ -263,18 +207,26 @@ def factorial(n):
...
+
+However, one can go too far with DRY and end up with difficult to understand code. You do not *need* code to be heavily abstracted wherever possible. The DRY principle should be used when the repetition becomes a problem, not when it "might become a problem".
 :::
 
 ## Style
 
-As mentioned [here](./Conducting-a-code-review.md#style), style is usually down to personal preference and it can sometimes be difficult to separate views on style from readability. Style should only be taken into account if it helps with consistency across the codebase.
+As mentioned [here](./Conducting-a-code-review.md#useful-style), matters of style usually comes down to the reviewers personal preference. Such comments on code style are classed as 'nits' and generally should be avoided in a code review. Nits usually distract developers away from the actual problems present in the codebase. Making comments about specific libraries used, variable name intracies *etc.* is not helpful to the reviewee (and usually discourages developers from contributing in the future).
```
[the remark this answers is not in the document]
Comment (sof202) [reply documentation/docs/Developer-information/Code-review/Code-examples.md]: I've collapsed all the sections on style into one now.

**Review by SiyiSEA — APPROVED** [review]
Well done

**Pull request merged**